iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -0,0 +1,160 @@
+    SmallVector<unsigned, 8> iteratorFolded;
+    SmallVector<Operation*, 8> resultOps;
+    auto terminator = cast<scf::YieldOp>(forOp.getBody()->getTerminator());
+    SmallVector<Value, 8> returnValues(terminator.getOperands().begin(),
+                                       terminator.getOperands().end());
+    SmallVector<Value, 8> initArgs(forOp.getIterOperands().begin(),
+                                   forOp.getIterOperands().end());
+    for (auto it : llvm::enumerate(forOp.getRegionIterArgs())) {
+      if (!it.value().hasOneUse()) continue;
+      Operation* op = it.value().use_begin()->getOwner();
+      if (!isa<vector::ShapeCastOp, vector::ExtractOp>(op)) continue;
+      Operation* returnValDef = returnValues[it.index()].getDefiningOp();
+      Value newReturn =
+          FoldCarryDep(forOp, op, returnValDef, initArgs[it.index()]);
+      if (newReturn) {
+        iteratorFolded.push_back(it.index());
+        resultOps.push_back(returnValDef);
+        returnValues[it.index()] = newReturn;
+
+        BlockAndValueMapping mapping;
+        mapping.map(op->getOperand(0), initArgs[it.index()]);
+        initArgs[it.index()] = rewriter.clone(*op, mapping)->getResult(0);
+      }
+    }
+    if (iteratorFolded.empty()) return failure();
+    auto newLoop =
+        rewriter.create<scf::ForOp>(forOp.getLoc(), forOp.lowerBound(),
+                                    forOp.upperBound(), forOp.step(), initArgs);
+    transferBody(forOp.getBody(), newLoop.getBody(), returnValues, rewriter);
+
+    // Replace the operation by the new one.
+    SmallVector<Value, 8> repResults(newLoop.getResults().begin(),
+                                     newLoop.getResults().end());
+    for (auto en : llvm::enumerate(iteratorFolded)) {
+      BlockAndValueMapping mapping;
+      mapping.map(returnValues[en.value()], newLoop.getResult(en.value()));
+      repResults[en.index()] =
+          rewriter.clone(*resultOps[en.index()], mapping)->getResult(0);
+      Operation* oldOp =
+          newLoop.getRegionIterArgs()[en.index()].use_begin()->getOwner();
+      SmallVector<Value, 1> arg(1, newLoop.getRegionIterArgs()[en.index()]);
+      oldOp->replaceAllUsesWith(arg);
+    }
+    auto a = forOp.getParentOp();
+    rewriter.replaceOp(forOp, repResults);
+    a->dump();
+    return success();
+  }
+};
+
+struct ForOpCanonicalizationPass
+    : PassWrapper<ForOpCanonicalizationPass, FunctionPass> {
+  void runOnFunction() override {
+    FuncOp fn = getFunction();
+    OwningRewritePatternList patterns;
+    patterns.insert<ForOpArgFolding>(fn.getContext());
+    applyPatternsAndFoldGreedily(fn, patterns);
+  }
+};
+}  // namespace
+
+std::unique_ptr<FunctionPass> createForOpCanonicalizationPass() {
+  return std::make_unique<ForOpCanonicalizationPass>();
+}
+
+static PassRegistration<ForOpCanonicalizationPass> pass(
+    "iree-codegen-forop-canonicalizatio-pass",
+    "Canonicalize for op loop carried dependency",
```

Comment:
"An ad-hoc pass to canonicalize selected loop carried dependencies on scf.for" to be more specific?